### PR TITLE
Added option to select the Target in Link Field

### LINKED ISSUE #16244: [LinkField] Add option to select the Target - _blank, _self, _parent and _top
### Is your feature request related to a problem? Please describe.
LinkField now only has URL and Text properties. However, having a Target property will help set the "target" attribute in the anchor tag HTML element.

### Describe the solution you'd like
Add Target Property in LinkField Class and expose a select list in the LinkField.Edit.cshtml to choose the target attribute.

### Describe alternatives you've considered
I tried inheriting the LinkField in my module and adding the Target property - Added the LinkField from my module and got error in a dictionary that the key already exists.

### PR DESCRIPTION
**Pull request by vengi83644**
Fixes #16244

image

image

### CODE REVIEW

**Comment by vengi83644** [comment]
https://github.com/OrchardCMS/OrchardCore/issues/16244

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs [hunk 1]
```diff
@@ -92,6 +93,7 @@ public override void Describe(DescribeContext<ContentItem> context)
                                 BigUrl = pair.Field.Url,
                                 Text = pair.Field.Text?[..Math.Min(pair.Field.Text.Length, LinkFieldIndex.MaxTextSize)],
                                 BigText = pair.Field.Text,
+                                Target = pair.Field.Target
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs:96]:
```suggestion
                                Target = pair.Field.Target,
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml [hunk 2]
```diff
@@ -45,5 +45,27 @@
...
         </div>
+        <div class="row">
+            <div class="col-md-6" asp-validation-class-for="Target">
+                <label asp-for="Target" class="@Orchard.GetLabelClasses()">
+                    @T["Target"]
+                </label>
+                <div class="@Orchard.GetEndClasses()">
+                    <input asp-for="Target" list="targetOptions" class="form-control" />
+                    <datalist id="targetOptions">
+                        <option value="_self">@T["Opens the linked document in the same frame as it was clicked (this is default)."]</option>
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml:56]:
```suggestion
                        <option value="_self">@T["Opens the linked document in the same frame as it was clicked (default)."]</option>
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml [hunk 3]
```diff
@@ -45,5 +45,27 @@
...
+                        <option value="_blank">@T["Opens the linked document in a new window or tab."]</option>
+                        <option value="_parent">@T["Opens the linked document in the parent frame."]</option>
+                        <option value="_top">@T["Opens the linked document in the full body of the window."]</option>
+                    </datalist>
+                    <span class="hint">
+                        @T["The target attribute of the A tag, see more:"]
+                        <a class="seedoc" target="_blank" href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target">
+                            target
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml:64]:
```suggestion
                            @T["Target"]
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkFieldSettings.Edit.cshtml [hunk 4]
```diff
@@ -57,3 +57,18 @@
...
+        <datalist id="targetOptions">
+            <option value="_self">@T["Opens the linked document in the same frame as it was clicked (this is default)."]</option>
+            <option value="_blank">@T["Opens the linked document in a new window or tab."]</option>
+            <option value="_parent">@T["Opens the linked document in the parent frame."]</option>
+            <option value="_top">@T["Opens the linked document in the full body of the window."]</option>
+        </datalist>
+        <span class="hint">@T["The default target proposed when creating a content item."]</span>
+        <span asp-validation-for="DefaultTarget" class="text-danger"></span>
+    </div>
+</div>
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkFieldSettings.Edit.cshtml:74]:
```suggestion
</div>

```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkFieldSettings.Edit.cshtml [hunk 5]
```diff
@@ -57,3 +57,18 @@
         <span asp-validation-for="DefaultText" class="text-danger"></span>
     </div>
 </div>
+
+<div class="row">
+    <div class="mb-3 col-md-6">
+        <label asp-for="DefaultTarget" class="form-label">@T["Default value of the Target"]</label>
+        <input asp-for="DefaultTarget" list="targetOptions" class="form-control" />
+        <datalist id="targetOptions">
+            <option value="_self">@T["Opens the linked document in the same frame as it was clicked (this is default)."]</option>
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkFieldSettings.Edit.cshtml:66]:
```suggestion
            <option value="_self">@T["Opens the linked document in the same frame as it was clicked (default)."]</option>
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.cshtml [hunk 6]
```diff
@@ -42,7 +42,9 @@
 
         href = href.ToUriComponents();
     }
+
+    var target = !string.IsNullOrWhiteSpace(Model.Target) ? Model.Target : "";
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.cshtml:46]:
```suggestion
    var target = !string.IsNullOrWhiteSpace(Model.Target) ? Model.Target : "_self";
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml [hunk 7]
```diff
@@ -45,5 +45,27 @@
...
+                <div class="@Orchard.GetEndClasses()">
+                    <input asp-for="Target" list="targetOptions" class="form-control" />
+                    <datalist id="targetOptions">
+                        <option value="_self">@T["Opens the linked document in the same frame as it was clicked (default)."]</option>
+                        <option value="_blank">@T["Opens the linked document in a new window or tab."]</option>
+                        <option value="_parent">@T["Opens the linked document in the parent frame."]</option>
+                        <option value="_top">@T["Opens the linked document in the full body of the window."]</option>
+                    </datalist>
+                    <span class="hint">
+                        @T["The target attribute of the A tag, see more:"]
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkField.Edit.cshtml:62]:
```suggestion
                        @T["The target attribute of the anchor tag, see more:"]
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@MikeAlhayek waiting for your approval or I will merge tonight

**Review by MikeAlhayek — APPROVED** [review]

**Comment by sebastienros** [comment]
Don't we need to update the migration if there is a new field in the index table?
Do we really need to index link targets? Is there any business need to query links by their html target?

**Comment by vengi83644** [comment]
We had a situation where we wanted to know how and where we used external links in our pages. One example page is this https://worldatwork.org/events

I missed the migrations. I will create a PR soon.

**Comment by sebastienros** [comment]
I feel like this is mixing business and ui logic. I would assume these kind of links are specific to a very special kind of entity it's describing, and the fact this external link is opening in a new window is just based on this flag. For instance, a "Vendor" could be "External" and based on that the link to the vendor should use a "_blank" link.

This not only would feel right (ok this is a personal opinion) but this would not require this "niche" feature (indexing UI values). Otherwise this opens us to accepting everything in the index.

**Comment by sebastienros** [comment]
@vengi83644 just to try to prove my point, where in your site, as a user, do I see the list of the things that are filtered by this new indexed value? Like is there a page of specific things that open in _blank, or a checkbox/filter for that?

**Comment by MikeAlhayek** [comment]
@sebastienros I think we should have the Target as settings but there is no need for indexing it.

**Comment by vengi83644** [comment]
@sebastienros @MikeAlhayek @Piedone We let our content authors get reports from the CMS for various use cases - examples: contents using certain widgets, certain content types with their title and description, contents of widgets used in different content types, a typical SEO report. Below is the screenshot of the page,
image

I am OK with removing the field from the index as it's not a use case for the larger audience. Thanks.

**Comment by Piedone** [comment]
OK, thanks, we're removing it, then. Please follow Sebastien's above recommendations instead.

**Pull request merged**